src/trio/_tests/test_testing_raisesgroup.py

```diff
@@ -0,0 +1,267 @@
+
+
+def test_RaisesGroup_matches() -> None:
+    rg = RaisesGroup(ValueError)
+    assert not rg.matches(None)
+    assert not rg.matches(ValueError())
+    assert rg.matches(ExceptionGroup("", (ValueError(),)))
+
+
+def test_message() -> None:
+    with pytest.raises(
+        AssertionError,
+        match=re.escape(
+            f"DID NOT RAISE any exception, expected ExceptionGroup({ValueError!r},)"
+        ),
+    ):
+        with RaisesGroup(ValueError):
+            ...
+    with pytest.raises(
+        AssertionError,
+        match=re.escape(
+            f"DID NOT RAISE any exception, expected ExceptionGroup(ExceptionGroup({ValueError!r},),)"
+        ),
+    ):
+        with RaisesGroup(RaisesGroup(ValueError)):
+            ...
+
+
+def test_matcher() -> None:
+    with pytest.raises(
+        ValueError, match="^You must specify at least one parameter to match on.$"
+    ):
+        Matcher()
+
+    with RaisesGroup(Matcher(ValueError)):
+        raise ExceptionGroup("", (ValueError(),))
+    with pytest.raises(ExceptionGroup):
+        with RaisesGroup(Matcher(TypeError)):
+            raise ExceptionGroup("", (ValueError(),))
+
+
+def test_matcher_match() -> None:
+    with RaisesGroup(Matcher(ValueError, "foo")):
+        raise ExceptionGroup("", (ValueError("foo"),))
+    with pytest.raises(ExceptionGroup):
+        with RaisesGroup(Matcher(ValueError, "foo")):
+            raise ExceptionGroup("", (ValueError("bar"),))
+
+    # Can be used without specifying the type
+    with RaisesGroup(Matcher(match="foo")):
+        raise ExceptionGroup("", (ValueError("foo"),))
+    with pytest.raises(ExceptionGroup):
+        with RaisesGroup(Matcher(match="foo")):
+            raise ExceptionGroup("", (ValueError("bar"),))
+
+
+def test_Matcher_check() -> None:
+    def check_oserror_and_errno_is_5(e: BaseException) -> bool:
+        return isinstance(e, OSError) and e.errno == 5
+
+    with RaisesGroup(Matcher(check=check_oserror_and_errno_is_5)):
+        raise ExceptionGroup("", (OSError(5, ""),))
+
+    # specifying exception_type narrows the parameter type to the callable
+    def check_errno_is_5(e: OSError) -> bool:
+        return e.errno == 5
+
+    with RaisesGroup(Matcher(OSError, check=check_errno_is_5)):
+        raise ExceptionGroup("", (OSError(5, ""),))
+
+    with pytest.raises(ExceptionGroup):
+        with RaisesGroup(Matcher(OSError, check=check_errno_is_5)):
+            raise ExceptionGroup("", (OSError(6, ""),))
+
+
+def test__ExceptionInfo(monkeypatch: pytest.MonkeyPatch) -> None:
+    monkeypatch.setattr(
+        trio.testing._raises_group,
+        "ExceptionInfo",
+        trio.testing._raises_group._ExceptionInfo,
+    )
+    with trio.testing.RaisesGroup(ValueError) as excinfo:
+        raise ExceptionGroup("", (ValueError("hello"),))
+    assert excinfo.type is ExceptionGroup
+    assert excinfo.value.exceptions[0].args == ("hello",)
+    assert isinstance(excinfo.tb, TracebackType)
+
+
+if TYPE_CHECKING:
```

Comment:
It'd be better to move these to the `type_tests` folder, which is also run under Pyright. Use bare `# type: ignore` so we check that it errors for both checkers.